mne/preprocessing/_optical_density.py

```diff
@@ -0,0 +1,46 @@
+# Authors: Robert Luke <mail@robertluke.net>
+#
+# License: BSD (3-clause)
+
+from ..io import BaseRaw
+from ..io.constants import FIFF
+from ..utils import _validate_type
+from ..io.pick import _picks_to_idx
+
+from numpy import mean, abs, log, where
+
+
+def optical_density(raw):
+    r"""Convert NIRS raw data to optical density.
+
+    Parameters
+    ----------
+    raw : instance of Raw
+        The raw data.
+
+    Returns
+    -------
+    raw : instance of Raw
+        The modified raw instance. Operates in place.
+
+    """
+    _validate_type(raw, BaseRaw, 'raw')
+    picks = _picks_to_idx(raw.info, 'fnirs_raw')
+    data_means = mean(raw.get_data(), 1)
+
+    # The devices measure light intensity. Negative light intensities should
+    # not occur. If they do it is likely due to hardware or movement issues.
+    # Set all negative values to abs(x), this also has the benefit of ensuring
+    # that the means are all greater than zero for the division below.
+    if len(where(raw._data[picks] <= 0)[0]) > 0:
+        # TODO How to throw warning without exiting tests
+        print("    Negative intensities encountered. Setting to abs(x)")
```

Comment:
We have a special warning function that deals with `logger` properly:
```
from ..utils import warn

warn('Negative ...')
```
then in tests
```
with pytest.warns(RuntimeWarning, match='...'):
    ...
```
Also we never use `print` but instead `logger.info`, in case you actually want to "print" something later

Reply:
Thanks, good to know about the info function.